File: kin_cpp/secondary_elements/bellcrank.cpp
#include "bellcrank.h"
#include "../primary_elements/node.h"
#include <optional>

Bellcrank::Bellcrank(const std::vector<Node*>& nodes, Node* pivot, const StaticVector<double, 3UL>& pivot_direction)
    : nodes(nodes), pivot(pivot), pivot_direction(pivot_direction), angle(0.0) {
}

void Bellcrank::rotate(double rotation_angle) {
    // Rotate each node about the pivot with the pivot direction
    for (Node* node : nodes) {
        node->reset();
        node->rotate(*pivot, false, std::make_optional(pivot_direction), std::make_optional(rotation_angle));
    }
    
    angle = rotation_angle;
}

const std::vector<Node*>& Bellcrank::getNodes() const {
    return nodes;
}

Node* Bellcrank::getPivot() const {
    return pivot;
}

const StaticVector<double, 3UL>& Bellcrank::getPivotDirection() const {
    return pivot_direction;
}

double Bellcrank::getAngle() const {
    return angle;
}

File: vehicle_model/suspension_model/suspension_elements/secondary_elements/bellcrank.py
from vehicle_model.suspension_model.suspension_elements.primary_elements.node import Node
from vehicle_model.assets.misc_math import rotation_matrix
from typing import Sequence, Tuple, Union
import numpy as np


class Bellcrank:
    """
    ## Bellcrank

    Bellcrank object

    Parameters
    ----------
    *nodes : Sequence[Node]
        All distinct pickup Nodes on bellcrank

        - First entry should connect to push/pull rod
        - Final entry should connect to inboard rod/shock

    pivot : Node
        Bellcrank pivot Node

    pivot_direction : Tuple[float, float, float]
        Unit vector representing Bellcrank pivot axis
    """
    def __init__(self, *nodes: Node, pivot: Node, pivot_direction: Tuple[float, float, float]) -> None:
        self.nodes: Sequence[Node] = nodes
        self.pivot = pivot
        self.pivot_direction = pivot_direction

        self.angle: float = 0

    def rotate(self, angle: float) -> None:
        """
        ## Rotate

        Rotates bellcrank

        Parameters
        ----------
        angle : float
            Angle of rotation in radians
        """
        for node in self.nodes:
            node.reset()
            node.rotate(origin=self.pivot, direction=self.pivot_direction, angle=angle)
        
        self.angle = angle